compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/Canonicalization.cpp

```diff
@@ -992,6 +992,64 @@ struct TransposeOpCanon final : OpRewritePattern<mlir::stablehlo::TransposeOp> {
   }
 };
 
+/// Check if a `t` is a tensor with zero extents.
+static std::optional<RankedTensorType> isZeroExtent(Type t) {
+  auto operandType = dyn_cast<RankedTensorType>(t);
+  if (operandType &&
+      llvm::any_of(operandType.getShape(), [](int64_t s) { return s == 0; })) {
+    return operandType;
+  }
+  return std::nullopt;
+}
+
+// Replace instances of zero extent tensors with empty tensors of the same
+// type.
+struct ZeroExtentTensorCanon final : RewritePattern {
+  ZeroExtentTensorCanon(MLIRContext *context, PatternBenefit benefit)
+      : RewritePattern(MatchAnyOpTypeTag(), benefit, context) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op->getLoc();
+
+    if (!dyn_cast<::mlir::stablehlo::StablehloDialect>(op->getDialect())) {
+      return rewriter.notifyMatchFailure(op, "not stablehlo");
+    }
+
+    // If the result is a zero-extent tensor, replace the whole op with an empty
+    // tensor.
+    for (auto result : op->getResults()) {
+      auto resultType = isZeroExtent(result.getType());
+      if (!resultType) {
```

Comment:
You should also check if it is dynamic as we would need to proxy the dynamic shapes through.

Reply:
Done, added the check to the helper function instead (since I'm guessing it needs to happen for both cases)